Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,298 @@
+  · case prec f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.prec ih1 ih2
+  · case rfind f' _ ih =>
+    apply RecursiveIn.rfind ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma TuringReducible.partrec_of_zero (f : ℕ →. ℕ) (fRecInZero : f ≤ᵀ fun _ => Part.some 0) :
+  Nat.Partrec f := by
+  induction fRecInZero
+  · exact Nat.Partrec.zero
+  · exact Nat.Partrec.succ
+  · exact Nat.Partrec.left
+  · exact Nat.Partrec.right
+  · case oracle g hg =>
+    simp at hg; rw [hg]
+    exact Nat.Partrec.zero
+  · case pair g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.pair ih1 ih2
+  · case comp g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.comp ih1 ih2
+  · case prec g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.prec ih1 ih2
+  · case rfind g _ ih =>
+    exact Nat.Partrec.rfind ih
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything (f : ℕ →. ℕ) : Nat.Partrec f ↔ ∀ g, f ≤ᵀ g :=
+  ⟨(·.turingReducible), (· _ |>.partrec_of_zero)⟩
+
+/--
+Proof that turing reducibility is reflexive.
+-/
+theorem TuringReducible.refl (f : ℕ →. ℕ) : f ≤ᵀ f := by
+  apply RecursiveIn.oracle; simp
+
+/--
+Instance declaring that `TuringReducible` is reflexive.
+-/
+instance : IsRefl (ℕ →. ℕ) TuringReducible :=
+  ⟨fun f => TuringReducible.refl f⟩
+
+/--
+Proof that turing reducibility is transitive.
+-/
+theorem TuringReducible.trans {f g h : ℕ →. ℕ} (hg : f ≤ᵀ g) (hh : g ≤ᵀ h) :
+  f ≤ᵀ h := by
+  induction hg
+  · exact RecursiveIn.zero
+  · exact RecursiveIn.succ
+  · exact RecursiveIn.left
+  · exact RecursiveIn.right
+  · case oracle f' fg =>
+    simp at fg; rw [fg]; exact hh
+  · case pair f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.pair ih1 ih2
+  · case comp f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.comp ih1 ih2
+  · case prec f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.prec ih1 ih2
+  · case rfind f' _ ih =>
+    apply RecursiveIn.rfind ih
+
+/--
+Instance declaring that `TuringReducible` is transitive.
+-/
+instance : IsTrans (ℕ →. ℕ) TuringReducible :=
+  ⟨@TuringReducible.trans⟩
+
+/--
+Instance declaring that `TuringReducible` is a preorder.
+-/
+instance : IsPreorder (ℕ →. ℕ) TuringReducible where
+  refl := TuringReducible.refl
+
+/--
+Proof that `TuringEquivalent` is an equivalence relation.
+-/
+theorem TuringEquivalent.equivalence : Equivalence TuringEquivalent :=
+  (AntisymmRel.setoid _ _).iseqv
+
+/--
+Proof that `TuringEquivalent` is reflexive.
+-/
+@[refl]
+theorem TuringEquivalent.refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  Equivalence.refl equivalence f
+
+/--
+Proof that `TuringEquivalent` is symmetric.
+-/
+@[symm]
+theorem TuringEquivalent.symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  Equivalence.symm equivalence h
+
+/--
+Proof that `TuringEquivalent` is transitive.
+-/
+@[trans]
+theorem TuringEquivalent.trans (f g h : ℕ →. ℕ) (h1 : f ≡ᵀ g) (h2 : g ≡ᵀ h) : f ≡ᵀ h :=
```

Comment:
You should use unicode subscripts `h₁`, typed as `h\1` (though, while this naming scheme is outside of my experience, I suspect that renaming f g h to something else might make the hypothesis names more clear. Then again, I would wait for another reviewer to check me on this.)